hubblestack/extmods/hubble_mods/curl.py

```diff
@@ -0,0 +1,238 @@
+# -*- encoding: utf-8 -*-
+"""
+Module curl
+=============================
+
+This module allows for querying against URLs
+
+Note that this module doesn't actually shell out to curl. Instead, it uses
+the requests library, primarily for performance concerns.
+
+Also note that this module doesn't support chaining from other modules.
+This is due to security concerns -- because Hubble can collect arbitrary data from
+a system, we don't want an attacker to be able to send that data to arbitrary
+endpoints.
+
+Audit Example:
+---------------
+check_unique_id:
+  description: 'curl check'
+  tag: 'ADOBE-01'
+  implementations:
+    - filter:
+        grains: 'G@osfinger:CentOS*Linux-7'
+      hubble_version: '>3 AND <7 AND <8'
+      module: curl
+      items:
+        - args:
+            url: "https://adobe.com/someurl"
+            function: GET
+            params:
+                key1: val1
+                key2: val2
+            data: "payload for post/put"
+            headers:
+                header1: val1
+                header2: val2
+            username: user
+            password: pwd
+            timeout: 9
+            decode_json: True
+
+          comparator:
+            type: "string"
+            match: "host*"
+            is_regex: true
+
+Note: You can use any comparator depending upon the kind of data a url is returning.
+
+FDG Example:
+------------
+main:
+  description: 'curl check'
+  module: curl
+    args:
+        url: "https://adobe.com/someurl"
+        function: GET
+        params:
+            key1: val1
+            key2: val2
+        data: "payload for post/put"
+        headers:
+            header1: val1
+            header2: val2
+        username: user
+        password: pwd
+        timeout: 9
+        decode_json: True
+
+Mandatory Params:
+    url
+"""
+import os
+import logging
+import requests
+
+import hubblestack.extmods.module_runner.runner_factory as runner_factory
+import hubblestack.extmods.module_runner.runner_utils as runner_utils
+from hubblestack.utils.hubble_error import HubbleCheckValidationError
+from salt.exceptions import CommandExecutionError
+
+log = logging.getLogger(__name__)
+
+
+def validate_params(block_id, block_dict, extra_args=None):
+    """
+    Validate all mandatory params required for this module
+
+    :param block_id:
+        id of the block
+    :param block_dict:
+        parameter for this module
+    :param extra_args:
+        Extra argument dictionary, (If any)
+        Example: {'caller': 'Audit'}
+
+    Raises:
+        AuditCheckValidationError: For any validation error
+    """
+    log.debug('Module: curl Start validating params for check-id: {0}'.format(block_id))
+
+    # fetch required param
+    function_name = runner_utils.get_param_for_module(block_id, block_dict, 'function')
```

Comment:
When function is not a mandatory param why are we checking it here ?

Reply:
Its value is checked, if it is apart from get/put/post